core/Convert.php

```diff
@@ -414,8 +416,29 @@ public static function raw2url($title) {
 	 * sequences including \r, \r\n, \n, or unicode newline characters
 	 * @param string $nl The newline sequence to normalise to. Defaults to that
 	 * specified by the current OS
+	 * @return string
 	 */
 	public static function nl2os($data, $nl = PHP_EOL) {
 		return preg_replace('~\R~u', $nl, $data);
 	}
+
+	/**
+	 * Encode a value into a string that can be used as part of a filename. 
+	 *
+	 * @param mixed $val Value to be encoded
+	 * @return string
+	 */
+	public static function base64url_encode($val) {
+		return rtrim(strtr(base64_encode(json_encode($val)), '+/', '~_'), '=');
+	}
+
+	/**
+	 * Decode a value that was encoded with Convert::base64url_encode.
+	 *
+	 * @param string $val Value to be decoded
+	 * @return mixed Original value
+	 */
+	public static function base64url_decode($val) {
+		return json_decode(base64_decode(str_pad(strtr($val, '~_', '+/'), strlen($val) % 4, '=', STR_PAD_RIGHT)));
```

Comment:
You're going to run into issues here where you base64url_encode an associative array, and it is returned as a stdObj. Maybe json_decode should have second parameter true?